ambry-cloud/src/main/java/com/github/ambry/cloud/azure/ShardedStorageClient.java

```diff
@@ -0,0 +1,258 @@
+   * @param cloudConfig {@link CloudConfig} object.
+   * @param azureCloudConfig {@link AzureCloudConfig} object.
+   * @param azureMetrics {@link AzureMetrics} object.
+   * @param blobLayoutStrategy {@link AzureBlobLayoutStrategy} object.
+   */
+  public ShardedStorageClient(CloudConfig cloudConfig, AzureCloudConfig azureCloudConfig, AzureMetrics azureMetrics,
+      AzureBlobLayoutStrategy blobLayoutStrategy) throws ReflectiveOperationException {
+    List<AzureCloudConfig.StorageAccountInfo> storageAccountInfoList = azureCloudConfig.azureStorageAccountInfo;
+
+    storageClientList = new ArrayList<>();
+    storageClientPartitionBoundaries = new ArrayList<>();
+    for (AzureCloudConfig.StorageAccountInfo storageAccountInfo : storageAccountInfoList) {
+      storageClientList.add(Utils.getObj(azureCloudConfig.azureStorageClientClass, cloudConfig, azureCloudConfig,
+          azureMetrics, blobLayoutStrategy, storageAccountInfo));
+      storageClientPartitionBoundaries.add(storageAccountInfo.getPartitionRangeStart());
+    }
+    storageClientPartitionBoundaries.add(Integer.MAX_VALUE);
+  }
+  /**
+   * Constructor for {@link ShardedStorageClient}.
+   * @param blobStorageAsyncClient {@link BlobServiceAsyncClient} object.
+   * @param blobBatchAsyncClient {@link BlobBatchAsyncClient} object.
+   * @param azureMetrics {@link AzureMetrics} object.
+   * @param blobLayoutStrategy {@link AzureBlobLayoutStrategy} object.
+   * @param azureCloudConfig {@link AzureCloudConfig} object.
+   */
+  public ShardedStorageClient(BlobServiceAsyncClient blobStorageAsyncClient, BlobBatchAsyncClient blobBatchAsyncClient,
+      AzureMetrics azureMetrics, AzureBlobLayoutStrategy blobLayoutStrategy, AzureCloudConfig azureCloudConfig) throws ReflectiveOperationException {
+    List<AzureCloudConfig.StorageAccountInfo> storageAccountInfoList = azureCloudConfig.azureStorageAccountInfo;
+    if (storageAccountInfoList.isEmpty()) {
+      throw new IllegalArgumentException("No storage account provided to ShardedStorageClient");
+    }
+    storageClientList = new ArrayList<>();
+    storageClientPartitionBoundaries = new ArrayList<>();
+    for (AzureCloudConfig.StorageAccountInfo storageAccountInfo : storageAccountInfoList) {
+      storageClientList.add(Utils.getObj(azureCloudConfig.azureStorageClientClass, blobStorageAsyncClient, blobBatchAsyncClient,
+          azureMetrics, blobLayoutStrategy, azureCloudConfig, storageAccountInfo));
+      storageClientPartitionBoundaries.add(storageAccountInfo.getPartitionRangeStart());
+    }
+    storageClientPartitionBoundaries.add(Integer.MAX_VALUE);
+  }
+
+  /**
+   * Finds the StorageClient that corresponds to Azure storage account holding Ambry metadata blobs.
+   * @return {@link StorageClient} that corresponds to Azure storage account holding Ambry metadata blobs.
+   */
+  private StorageClient getMetadataAzureStorageClient() {
+    return storageClientList.get(0);
```

Comment:
I have concerns about this logic. What happens if somehow the ordering of accounts in the configs gets messed up (maybe when we are creating new config after account exapansion)?

I think a better logic would be to say that account responsible for partitionId "0" is the account responsible for metadata.

Reply:
AzureCloudConfig.parseStorageAccountInfo() ensures that that never happens